### PR TITLE
PyTorch Lightning example

### PR DESCRIPTION
**Pull request by ordabayevy**
This example shows how to train Pyro models using PyTorch Lightning and is adapted from [Horovod example](https://pyro.ai/examples/svi_horovod.html).

### CODE REVIEW

**Comment by ordabayevy** [comment]
Addresses #3171.

File: examples/svi_lightning.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+# We define an ELBO loss, a PyTorch optimizer, and a training step in our PyroLightningModule.
+# Note that we are using a PyTorch optimizer instead of a Pyro optimizer and
+# we are using ``training_step`` instead of Pyro's SVI machinery.
+class PyroLightningModule(pl.LightningModule):
+    def __init__(self, model, guide, lr):
+        super().__init__()
+        self.pyro_model = model
+        self.pyro_guide = guide
+        self.loss_fn = Trace_ELBO().differentiable_loss
```
Comment (eb8680) [inline examples/svi_lightning.py:58]:
```suggestion
    def __init__(self, loss_fn: pyro.infer.elbo.ELBOModule, lr: float):
        super().__init__()
        self.loss_fn = loss_fn
        self.model = loss_fn.model
        self.guide = loss_fn.guide
```

File: examples/svi_lightning.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        self.lr = lr
+
+    def training_step(self, batch, batch_idx):
+        """Training step for Pyro training."""
+        loss = self.loss_fn(self.pyro_model, self.pyro_guide, *batch)
```
Comment (eb8680) [inline examples/svi_lightning.py:63]:
```suggestion
        loss = self.loss_fn(*batch)
```

File: examples/svi_lightning.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        # Logging to TensorBoard by default
+        self.log("train_loss", loss)
+        return loss
+
+    def configure_optimizers(self):
+        """Configure an optimizer."""
+        return torch.optim.Adam(self.pyro_guide.parameters(), lr=self.lr)
```
Comment (eb8680) [inline examples/svi_lightning.py:70]:
```suggestion
        return torch.optim.Adam(self.loss_fn.parameters(), lr=self.lr)
```

File: examples/svi_lightning.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+
+def main(args):
+    # Create a model, synthetic data, a guide, and a lightning module.
+    pyro.set_rng_seed(args.seed)
+    model = Model(args.size)
+    covariates = torch.randn(args.size)
+    data = model(covariates)
+    guide = AutoNormal(model)
+    training_plan = PyroLightningModule(model, guide, args.learning_rate)
```
[the remark this answers is not in the document]
Comment (ordabayevy) [reply examples/svi_lightning.py]: I didn't know about `ELBOModule`. This is much neater!

**Review by ordabayevy — COMMENTED** [review]
Thanks for reviewing @eb8680. I think it is much neater now using `ELBOModule`!

**Review by fritzo — APPROVED** [review]
Looks great!  Can you just confirm the generated docs are readable, i.e. after running `make tutorial`?

**Comment by ordabayevy** [comment]
@fritzo There is something wrong with building tutorials when I run `make tutorial`:

```
make tutorial
make -C tutorial html
make[1]: Entering directory '/mnt/disks/dev/repos/pyro/tutorial'
Running Sphinx v6.1.3
building [mo]: targets for 0 po files that are out of date
writing output... 
building [html]: targets for 80 source files that are out of date
updating environment: [new config] 80 added, 0 changed, 0 removed
reading sources... [100%] svi_part_ii .. working_memory                                                                                                             

Warning, treated as error:
/mnt/disks/dev/repos/pyro/tutorial/source/gp.ipynb:973:Duplicate substitution definition name: "image0".
make[1]: *** [Makefile:20: html] Error 2
make[1]: Leaving directory '/mnt/disks/dev/repos/pyro/tutorial'
make: *** [Makefile:18: tutorial] Error 2
```
Trying to figure out what is wrong ... (if you know a quick fix would appreciate it)

**Comment by fritzo** [comment]
@ordabayevy not sure what's causing the build issue...

Unrelated, I see

> .../pyro-ppl/pyro/tutorial/source/svi_lightning.rst: WARNING: document isn't included in any toctree

Could you add svi_lightning to tutorial/source/index.rst so it shows up on the website?

**Comment by ordabayevy** [comment]
Still no luck with `make tutorial`. When I try to build tutorials on dev branch I get this:

<details>

```
make -C tutorial html
make[1]: Entering directory '/mnt/disks/dev/repos/pyro/tutorial'
Running Sphinx v6.1.3
making output directory... done
building [mo]: targets for 0 po files that are out of date
writing output... 
building [html]: targets for 79 source files that are out of date
updating environment: [new config] 79 added, 0 changed, 0 removed
reading sources... [100%] tensor_shapes .. working_memory                                                                                                           

Warning, treated as error:
/mnt/disks/dev/repos/pyro/tutorial/source/logistic-growth.ipynb:1220:File not found: 'workflow.html'
make[1]: *** [Makefile:20: html] Error 2
make[1]: Leaving directory '/mnt/disks/dev/repos/pyro/tutorial'
make: *** [Makefile:18: tutorial] Error 2
```

</details>

**Comment by ordabayevy** [comment]
> Can you just confirm the generated docs are readable, i.e. after running `make tutorial`? Also ensure the title isn't too long when it appears on the left hand side TOC.

I was able to build the tutorial by ignoring warnings and can confirm that the generated doc is readable and the title in the left hand side TOC is not too long.

**Review by fritzo — APPROVED** [review]
Looks great, thanks for building tutorials. I'll look into fixing those warnings.

@eb8680 any further comments?  I'll hold off merging, feel free to merge

**Review by eb8680 — APPROVED** [review]
LGTM

**Comment by ordabayevy** [comment]
Thanks @eb8680 and @fritzo for reviewing!

**Pull request merged**